src/Umbraco.AuthorizedServices/Services/Implement/DatabaseKeyStorage.cs

```diff
@@ -0,0 +1,88 @@
+using Microsoft.Extensions.Logging;
+using Umbraco.AuthorizedServices.Models;
+using Umbraco.AuthorizedServices.Persistence.Dtos;
+using Umbraco.Cms.Infrastructure.Scoping;
+
+namespace Umbraco.AuthorizedServices.Services.Implement;
+
+/// <summary>
+/// Implements <see cref="IKeyStorage"/> for API key storage using a database table.
+/// </summary>
+internal sealed class DatabaseKeyStorage : IKeyStorage
```

Comment:
Is it worth introducing a base class for both database key and token storage, e,g. `DatabaseAuthorizationParameterStorageBase`?  Might be you could share some code between this and `DatabaseTokenStorage`.  Not sure... but maybe have a look and see if you think it's worthwhile.

Reply:
It would make sense, applied the recommendation.